File: leetcode/python/0160_intersection_of_two_linked_lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        return self.alternatingPointers(headA, headB)
        # return self.twoPointersLength(headA, headB)
        # return self.seenSet(headA, headB)

    def alternatingPointers(self, headA, headB):
        first, second = headA, headB

        while first != second:
            first = headB if not first else first.next
            second = headA if not second else second.next
        
        return first
    
    def twoPointersLength(self, headA, headB):
        length1, length2 = self.length(headA), self.length(headB)
        if length1 > length2: headA, headB = headB, headA

        for _ in range(abs(length1 - length2)): headB = headB.next

        while headA and headB and headA != headB: headA, headB = headA.next, headB.next

        return headA
    
    def length(self, node):
        res = 0
        while node:
            res += 1
            node = node.next
        return res
    
    def seenSet(self, headA, headB):
        seen = set()
        while headA:
            seen.add(headA)
            headA = headA.next
        
        while headB:
            if headB in seen: return headB
            headB = headB.next
        
        return None

File: leetcode/cpp/0160_intersection_of_two_linked_lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
            set<ListNode*> setA, setB;

        for (auto tmpA = headA, tmpB = headB; tmpA != NULL || tmpB != NULL;) {
            if (setA.find(tmpB) == setA.end()) {
                setA.insert(tmpA);
            }
            else if (setA.find(tmpB) != setA.end()) {
                return tmpB;
            }

            if (setB.find(tmpA) == setB.end()) {
                setB.insert(tmpB);
            }
            else if (setB.find(tmpA) != setB.end()) {
                return tmpA;
            }
            
            if (tmpA == tmpB) {
                return tmpA;
            }

            if (tmpA != NULL) {
                tmpA = tmpA->next;
            }

            if (tmpB != NULL) {
                tmpB = tmpB->next;
            }
        }

        return NULL;
    }
};